### ISSUE TITLE
R/dt2vw.R error when used with many columns

### ISSUE THREAD

**Issue by martinbel** [JohnLangford/vowpal_wabbit#761]
The dt2vw function returns this error: "sprintf: only 100 arguments allowed" 
when called with a data.table of +100 columns. 

Is there any other way to sort this out? I've checked the source code, but couldn't find a solution. This would be quite useful to have working as can improve significantly the VW formating. 

Also, am I using the weight argument correctly?

```
# Here is a gist with the same function. Just added a print of the format string.

library(data.table)
source('https://gist.githubusercontent.com/martinbel/a013c2bd2333b85eeb76/raw/7b68ede0385f3d8684a07854e8d6ca527b9cd441/dt2vw.R')

# Generate a table with 101 columns
X = cbind(matrix(rnorm(1:1010), nrow=10, ncol=101), sample(c(1, -1), 10, T))
X = as.data.table(X)
X[, weight:=ifelse(V102 == 1, 3, 1)]

namespaces = list(a = list(varName = sprintf('V%s',1:50), keepSpace=F),
                  b = list(varName = sprintf('V%s',51:101), keepSpace=F))

dt2vw(data=X, fileName='../out.vw', namespaces = namespaces, 
      target='V102', weight = 'weight', 
      tag = NULL, hard_parse = F)

### error

[1] "sprintf('%f %f |a V1:%f V2:%f V3:%f V4:%f V5:%f V6:%f V7:%f V8:%f V9:%f V10:%f V11:%f V12:%f V13:%f V14:%f V15:%f V16:%f V17:%f V18:%f V19:%f V20:%f V21:%f V22:%f V23:%f V24:%f V25:%f V26:%f V27:%f V28:%f V29:%f V30:%f V31:%f V32:%f V33:%f V34:%f V35:%f V36:%f V37:%f V38:%f V39:%f V40:%f V41:%f V42:%f V43:%f V44:%f V45:%f V46:%f V47:%f V48:%f V49:%f V50:%f  |b V51:%f V52:%f V53:%f V54:%f V55:%f V56:%f V57:%f V58:%f V59:%f V60:%f V61:%f V62:%f V63:%f V64:%f V65:%f V66:%f V67:%f V68:%f V69:%f V70:%f V71:%f V72:%f V73:%f V74:%f V75:%f V76:%f V77:%f V78:%f V79:%f V80:%f V81:%f V82:%f V83:%f V84:%f V85:%f V86:%f V87:%f V88:%f V89:%f V90:%f V91:%f V92:%f V93:%f V94:%f V95:%f V96:%f V97:%f V98:%f V99:%f V100:%f V101:%f ',V102, weight, V1, V2, V3, V4, V5, V6, V7, V8, V9, V10, V11, V12, V13, V14, V15, V16, V17, V18, V19, V20, V21, V22, V23, V24, V25, V26, V27, V28, V29, V30, V31, V32, V33, V34, V35, V36, V37, V38, V39, V40, V41, V42, V43, V44, V45, V46, V47, V48, V49, V50, V51, V52, V53, V54, V55, V56, V57, V58, V59, V60, V61, V62, V63, V64, V65, V66, V67, V68, V69, V70, V71, V72, V73, V74, V75, V76, V77, V78, V79, V80, V81, V82, V83, V84, V85, V86, V87, V88, V89, V90, V91, V92, V93, V94, V95, V96, V97, V98, V99, V100, V101)"
Error in sprintf("%f %f |a V1:%f V2:%f V3:%f V4:%f V5:%f V6:%f V7:%f V8:%f V9:%f V10:%f V11:%f V12:%f V13:%f V14:%f V15:%f V16:%f V17:%f V18:%f V19:%f V20:%f V21:%f V22:%f V23:%f V24:%f V25:%f V26:%f V27:%f V28:%f V29:%f V30:%f V31:%f V32:%f V33:%f V34:%f V35:%f V36:%f V37:%f V38:%f V39:%f V40:%f V41:%f V42:%f V43:%f V44:%f V45:%f V46:%f V47:%f V48:%f V49:%f V50:%f  |b V51:%f V52:%f V53:%f V54:%f V55:%f V56:%f V57:%f V58:%f V59:%f V60:%f V61:%f V62:%f V63:%f V64:%f V65:%f V66:%f V67:%f V68:%f V69:%f V70:%f V71:%f V72:%f V73:%f V74:%f V75:%f V76:%f V77:%f V78:%f V79:%f V80:%f V81:%f V82:%f V83:%f V84:%f V85:%f V86:%f V87:%f V88:%f V89:%f V90:%f V91:%f V92:%f V93:%f V94:%f V95:%f V96:%f V97:%f V98:%f V99:%f V100:%f V101:%f ",  : 
  only 100 arguments are allowed
Called from: sprintf("%f %f |a V1:%f V2:%f V3:%f V4:%f V5:%f V6:%f V7:%f V8:%f V9:%f V10:%f V11:%f V12:%f V13:%f V14:%f V15:%f V16:%f V17:%f V18:%f V19:%f V20:%f V21:%f V22:%f V23:%f V24:%f V25:%f V26:%f V27:%f V28:%f V29:%f V30:%f V31:%f V32:%f V33:%f V34:%f V35:%f V36:%f V37:%f V38:%f V39:%f V40:%f V41:%f V42:%f V43:%f V44:%f V45:%f V46:%f V47:%f V48:%f V49:%f V50:%f  |b V51:%f V52:%f V53:%f V54:%f V55:%f V56:%f V57:%f V58:%f V59:%f V60:%f V61:%f V62:%f V63:%f V64:%f V65:%f V66:%f V67:%f V68:%f V69:%f V70:%f V71:%f V72:%f V73:%f V74:%f V75:%f V76:%f V77:%f V78:%f V79:%f V80:%f V81:%f V82:%f V83:%f V84:%f V85:%f V86:%f V87:%f V88:%f V89:%f V90:%f V91:%f V92:%f V93:%f V94:%f V95:%f V96:%f V97:%f V98:%f V99:%f V100:%f V101:%f ", 
    V102, weight, V1, V2, V3, V4, V5, V6, V7, V8, V9, V10, V11, 
    V12, V13, V14, V15, V16, V17, V18, V19, V20, V21, V22, V23, 
    V24, V25, V26, V27, V28, V29, V30, V31, V32, V33, V34, V35, 
    V36, V37, V38, V39, V40, V41, V42, V43, V44, V45, V46, V47, 
    V48, V49, V50, V51, V52, V53, V54, V55, V56, V57, V58, V59, 
    V60, V61, V62, V63, V64, V65, V66, V67, V68, V69, V70, V71, 
    V72, V73, V74, V75, V76, V77, V78, V79, V80, V81, V82, V83, 
    V84, V85, V86, V87, V88, V89, V90, V91, V92, V93, V94, V95, 
    V96, V97, V98, V99, V100, V101)

# works with less columns
dt2vw(data=X[, 50:dim(X)[2], with=FALSE], 
      fileName='../out.vw', namespaces = NULL,  #set to null as there are less columns 
      target='V102', weight = 'weight', 
      tag = NULL, hard_parse = F)

system('head -3 ../out.vw')
1.000000 3.000000 |A V50:-0.417908 V51:-0.524908 V52:-0.873840 V53:-0.893032 V54:0.626959 V55:-0.332472 V56:1.191455 V57:0.116012 V58:0.497325 V59:-0.310254 V60:-0.588995 V61:0.379087 V62:0.361883 V63:-0.397330 V64:1.806091 V65:0.669346 V66:0.563994 V67:0.371265 V68:-0.822683 V69:-0.366743 V70:-0.064391 V71:-1.215438 V72:-0.386661 V73:-0.226746 V74:2.193334 V75:0.371084 V76:-1.108630 V77:-0.939769 V78:2.420315 V79:0.158951 V80:0.619462 V81:-0.420737 V82:-0.104049 V83:1.487708 V84:0.490672 V85:0.486206 V86:-0.661062 V87:-0.421320 V88:-0.882243 V89:-1.015385 V90:0.709694 V91:0.234034 V92:0.623823 V93:-0.137862 V94:0.302321 V95:0.208074 V96:0.086683 V97:0.212727 V98:0.425434 V99:-1.068463 V100:-0.530487 V101:-0.945378 
-1.000000 1.000000 |A V50:0.962751 V51:1.582858 V52:0.752361 V53:-1.130187 V54:0.732169 V55:-0.557810 V56:0.737294 V57:1.065086 V58:0.526012 V59:1.138539 V60:0.396420 V61:-0.137370 V62:0.112700 V63:0.886309 V64:-1.015369 V65:1.839683 V66:-0.580796 V67:1.378569 V68:-2.650571 V69:0.461465 V70:2.086908 V71:-0.821210 V72:-1.431688 V73:1.439760 V74:0.228630 V75:-1.972637 V76:-0.540819 V77:-0.996993 V78:-0.150147 V79:-1.262729 V80:0.599682 V81:2.085684 V82:-0.620893 V83:-0.918864 V84:-0.501955 V85:2.457091 V86:0.215191 V87:-0.878072 V88:-0.546416 V89:1.376766 V90:0.161631 V91:-0.731223 V92:0.282834 V93:2.214595 V94:0.061333 V95:-0.324394 V96:-0.053773 V97:0.019203 V98:1.098778 V99:-0.420786 V100:0.013317 V101:-0.345533 
```

**Comment by arielf** [comment]
@martinbel 

Thanks for the report.

2 comments:

1) The arbitrary 100 params limit in R/sprintf is unfortunate.  It is not a vowpal-wabbit issue as much as it is an R fundamental issue.  The following script demonstrates it:
```
#!/usr/bin/Rscript --vanilla
#
# Demonstrate 100-char limitation in R sprintf

# Works with N up to 99 (format string is additional arg).
# Breaks beyond 99 with:
#       Error in (function (fmt, ...)  : only 100 arguments are allowed

N <- 100
fmt.arg <- paste(rep('%d ', N), sep='', collapse='')
vec <- 1:N
do.call(sprintf, c(list(fmt.arg), vec))

# Also: using 'cat' instead of 'sprintf' in the 'do.call' works fine too, so this seems like a (R) sprintf specific limitation
```

2) To get more traction or hopefully a work-around, I would try to email the contributors/fixers of this code:

See:
```
     git log R/dt2vw.R
```
for their email addresses.

**Comment by arielf** [comment]
@martinbel
To your other question: I see both 1.000000 and 3.000000 weights in the partial output you provided, so Imy impression is your use of weight is correct.

**Comment by martinbel** [comment]
@slymore you are the function author, i believe? Any ideas how to make it work with +100 colums?
I've asked in R-devel if it's possible to lift this restriction, let's see how that goes...

**Comment by martinbel** [comment]
@arielf Here is an attempt to solve some of the issues. It works with more columns but it's far from memory efficient (has some data duplication in there). Anyway, it's simpler and can be easily modified. 
Would you mind giving trying it, in order to see if it's the correct input format? 
The example just generates two namespaces for potential categorical/numeric features. Also, the zeros are removed. This is what i'm not sure is entirely correct.


```
#### Example
# Parameters are similar to dt2vw
# rm(list = ls(all = TRUE)); gc()
#
# library(data.table)
#
# X = cbind(matrix(rpois(1010, 1), nrow=10, ncol=1001), sample(c(1, -1), 10, T))
# X = as.data.table(X)
# X[, weight:=ifelse(V102 == 1, 3, 1)]
# X[, y_train:=V102]
# X[, V102:=NULL]
#
# namespaces = list(n = list(varName = sprintf('V%s',1:50), keepSpace=F, dtype='numeric'),
#                   c = list(varName = sprintf('V%s',51:101), keepSpace=F, dtype='categorical'))
# data.table.vw(data=X, fileName='../out.vw', namespaces = namespaces,
#               target='y_train', weight = 'weight',
#               tag = NULL, hard_parse = F)

data.table.vw <- function(data=X, fileName='../out.vw', namespaces = namespaces,
                          target='y_train', weight = 'weight',
                          tag = NULL, hard_parse = F){

  #change target if its boolean to take values in {-1,1}
  if(is.logical(data[[target]]) | sum(levels(factor(data[[target]])) == levels(factor(c(0,1)))) == 2){
    data[[target]][data[[target]] == TRUE] = 1
    data[[target]][data[[target]] == FALSE] = -1
  }

  #if namespaces = NULL, define a unique namespace
  if(is.null(namespaces)){
    all_vars = colnames(data)[!colnames(data) %in% c(target, weight, tag)]
    namespaces <- list(A = list(varName = all_vars, keepSpace=F))
  }

  #parse variable names
  specChar = '\\(|\\)|\\||\\:'
  specCharSpace = '\\(|\\)|\\||\\:| '

  parsingNames <- function(x){
    ret = c()
    for(el in x)
      ret = append(ret, gsub(specCharSpace,'_', el))
    ret
  }

  #parse categorical variables
  parsingVar <- function(x, keepSpace, hard_parse){
    #remove leading and trailing spaces, then remove special characters then remove isolated underscores.
    if(!keepSpace)
      spch = specCharSpace
    else
      spch = specChar

    if(hard_parse)
      gsub('(^_( *|_*)+)|(^_$)|(( *|_*)+_$)|( +_+ +)',' ', gsub(specChar,'_', gsub('(^ +)|( +$)', '',x)))
    else
      gsub(spch, '_', x)
  }

  ###   NAMESPACE LOAD WITH A YAML FILE
  if(typeof(namespaces) == "character" && length(namespaces) == 1 && str_sub(namespaces, -4, -1) == "yaml"){
    print("###############  USING YAML FILE FOR LOADING THE NAMESPACES  ###############")
    library(yaml)
    namespaces = yaml.load_file(namespaces)
  }

  ###   AVOIDING DATA FORMAT PROBLEMS
  setnames(data, names(data), parsingNames(names(data)))
  names(namespaces) <- parsingNames(names(namespaces))
  for(x in names(namespaces)) namespaces[[x]]$varName = parsingNames(namespaces[[x]]$varName)
  target = parsingNames(target)
  if(!is.null(tag)) tag = parsingNames(tag)
  if(!is.null(weight)) weight = parsingNames(weight)


  # n = namespaces[[1]]
  # j = n[['varName']][[1]]
  # col = data[[j]]

  # # Just in case, test this
  # if(class(col) == 'numeric'){
  #   formater = '%s:%f'
  # } else {
  #   formater = '%s_%s'
  # }

  format_vw <- function(column, formater){
    ifelse(as.logical(column), sprintf(formater, j, column), '')
  }

  for(n in namespaces){
    if(n[['dtype']] == 'numeric'){
      for(j in n[['varName']]){
        column = data[[j]]
        formater = '%s:%f'
        set(data, i=NULL, j=j, value=format_vw(column, formater))
        # Just in case, test this
        #      if(class(col) == 'numeric'){
        #        formater = '%s:%f'
        #      }
        #else {
        #  formater = '%s_%s'
        #}

      }
    } else {
      for(j in n[['varName']]){
        formater = '%s_%s'
        set(data, i=NULL, j=j, value=format_vw(data[[j]], formater))
      }
    }
  }

  ###   FIRST PART OF THE VW DATA FORMAT: target, weight, tag
  formatDataVW = ''
  argexpr = character(0)

  ### Label can be null, no training is performed
  if(!is.null(target)){
    # Both weight and tag are not null
    if(!is.null(weight) && !is.null(tag)){
      formatDataVW = '%f %f %s'
      argexpr = paste(target, weight, tag, sep = ', ')
    }
    # Weight is null, tag is not null
    else if(is.null(weight) && !is.null(tag)){
      formatDataVW = '%f %s'
      argexpr = paste(target, tag, sep = ', ')
    }
    # Weight is not null, tag is null
    else if(!is.null(weight) && is.null(tag)){
      formatDataVW = '%f %f'
      argexpr = paste(target, weight, sep = ', ')
    }
    # We just output target
    else{
      formatDataVW = '%f'
      argexpr = target
    }
  }

  formatDataVW = paste0("sprintf('", formatDataVW, "',",argexpr, ")")
  y = data[, eval(parse(text = formatDataVW))]

  nmp2data <- function(x, data){
    cbind(sprintf('|%s', x), data[, namespaces[[x]][['varName']], with=FALSE])
  }

  nmp_data = lapply(names(namespaces), function(x) nmp2data(x, data))
  data = cbind(y, do.call(cbind, nmp_data))

  drop_extra_space <- function(x){
    gsub(' {1,}', ' ', x)
  }

  out = apply(data, 1, function(x) drop_extra_space(paste(x, collapse=' ')))

  write.table(out, file=fileName, sep=' ', row.names=FALSE,
              col.names = FALSE, quote=FALSE)

}
```

**Comment by slymore** [comment]
Hi @martinbel ,

Thanks for the report. Indeed ```sprintf```  limits the number of arguments to 100.
In dt2vw the full formatting string is created then evaluated inside the data table, which avoids any data replications.
```
temp = data[, eval(parse(text = formatDataVW))]
```

As a workaround, I would suggest trying to improve on the existing code than rewriting another version. As is, your version works but there are several problems:
- you have to put all the numerical variables in a single namespace, same for categorical.
```  if(n[['dtype']] == 'numeric') ...```
 In a real dataset you might not want to determine which variables are numerical / categorical beforehand. Also you might want to put in a same namespace numerical and categorical variables
- It's much slower than dt2vw. 5 times slower on this quick test code:
```
library(rbenchmark)
N = 100000
ncls = 90
dt = data.table(rep(1,N))
lvls = c("tea", "coffee", "chocolate")
for(i in 2:ncls)
{
  tt = data.table(sample(lvls, N, replace = T))
  setnames(tt, colnames(tt), paste("V",i, sep = ""))
  dt = cbind(dt, tt)
}

namespaces = list(n = list(varName = sprintf('V%s',2:ncls), keepSpace=F, dtype='numeric'))

rbenchmark::benchmark(dt2vw = dt2vw(data = dt, fileName = "~/Desktop/test.dt2vw",namespaces = namespaces, target = "V1",weight = NULL, hard_parse = F),
                      alt = data.table.vw(data=dt, fileName='~/Desktop/alt.vw', namespaces = namespaces,target='V1', weight = NULL,tag = NULL, hard_parse = F),
                      replications = 1)
```

In my opinion sticking with a ```sprintf``` version of dt2vw is better as it avoids all data replications.
We might be able to overcome the 100 argument limit by splitting the formatting string into substrings, will look into that.

That said, dt2vw would definitely gain in being more readable and is far from perfect. Feel free to ask about some portions of code where you don't understand what it's doing. Any feedback is welcome

**Comment by martinbel** [comment]
1) I see your point. This was more a hack than a real contribution, it's slower but actually works with more columns. It's unfortunate sprintf is so limited, but given this limitation I don't see an easy way to modify your function, if you can change that, it would be great.
2) Also, another interesting adition would be to allow the function to work with chunks of data. Unfortunately fread has some issues with this, but should be solved eventually.
3) The point of adding a "dtype" to the namespace is that sometimes the categorical features are already encoded as numbers, so just checking it's class won't work. (I'm not completely sure this is what you are doint in R/dt2vw.R)
A better solution to this would be to allow the user to pass the feature data types to the namespaces, or determine them in a cleverer way. But I think it's better the user determines which features should be treated as categorical or numerical. Maybe just adding as agument a data.table with "variable_name | data type" as input could be an idea, so it's not constrained to namespaces.
If I make a more  decent improvement i'll contact you again. Thanks for your comments.

**Comment by martinbel** [comment]
R-devel got back with this hack for sprintf. I've tested it in dt2vw and works ok. This means however adding a new function to the dt2vw.R source code and changing sprintf for sprintf2.

```
## Work around the "only 100 arguments are allowed" error
## in base::sprintf(). Only works with 'fmt' of length 1.
sprintf2 <- function(fmt, ...)
{
    MAX_NVAL <- 99L
    args <- list(...)
    if (length(args) <= MAX_NVAL)
        return(sprintf(fmt, ...))
    stopifnot(length(fmt) == 1L)
    not_a_spec_at <- gregexpr("%%", fmt, fixed=TRUE)[[1L]]
    not_a_spec_at <- c(not_a_spec_at, not_a_spec_at + 1L)
    spec_at <- setdiff(gregexpr("%", fmt, fixed=TRUE)[[1L]], not_a_spec_at)
    nspec <- length(spec_at)
    if (length(args) < nspec)
        stop("too few arguments")
    if (nspec <= MAX_NVAL) {
        break_points <- integer(0)
    } else {
        break_points <- seq(MAX_NVAL + 1L, nspec, by=MAX_NVAL)
    }
    break_from <- c(1L, break_points)
    break_to <- c(break_points - 1L, nspec)
    fmt_break_at <- spec_at[break_points]
    fmt_chunks <- substr(rep.int(fmt, length(fmt_break_at) + 1L),
                         c(1L, fmt_break_at),
                         c(fmt_break_at - 1L, nchar(fmt)))
    ans_chunks <- mapply(
        function(fmt_chunk, from, to)
            do.call(sprintf, c(list(fmt_chunk), args[from:to])),
        fmt_chunks,
        break_from,
        break_to
    )
    paste(ans_chunks, collapse="")
}
```

**Comment by SelimRaboudi** [comment]
Nice ! I put the code for sprintf2 in dt2vw.
(the only modification for it to work is to replace the last line by ```paste(apply(ans_chunks,1, paste, collapse = ""), collapse = "\n")```)

1) Seems it's solved thanks to your contribution. I will do a PR with the new code


2) what exactly do you mean by " chunks of data" ? The input for dt2vw is a data table / data frame. You can get a chunk of data beforehand using fread or another function as you said, then transform it to a data table and feed it to dt2vw ( It seems more logical to create the desired chunk outside dt2vw )


3) In the current version dt2vw indeed checks for the class of each column.
```
#Index: check if the variable is numerical (->TRUE) or categorical (->FALSE)
Index = list()
  
  for(namespaceName in names(namespaces))
  {
    Index[[namespaceName]] = sapply(data[,namespaces[[namespaceName]][['varName']],with=F], is.numeric)
```
Yet I think to handle this case it would be easier to transform the column into a categorical one before giving the dt to dt2vw.
Ex: a variable "class" (1 / 2/ 3) --> (Class_1 / Class_2 / Class_3)
dt2vw would have to do this operation anyways for your variable to be correctly interpreted by VW.

adding "variable_name | data type" argument: I think it would be cumbersome from a user point of vue. in most cases it just forces you to create another (unnecessary) argument before running dt2vw.

I see your point though, will think about other alternatives

**Comment by martinbel** [comment]
Great! Glad to have helped! It's R-devel merit anyway! I think having a great dt2vw function can make more people using R closer to vw. As most of the preprocessing scripts are in python, this would help. 
I'm also using an R script to do a parameter search using perf, I'll make a commit with this soon.

(3) Makes sense. I guess if you have categorical variables encoded as numbers, you can coerce them to factor beforehand, so then dt2vw would handle them as categorical. True, no need to add anything. Totally agree with your comments.

(2) With "reading the data in chunks" I mean that you could read the file in parts and then be exported in pieces in vw format. Normally a great motivation for using vw is that your data doesn't fit in memory. So I believe this should be useful. Also it should be quite fast (compared to other iterative options), because for each chunk it's vectorized.
So this way, you would be able to read a file in parts (chunks) and write it in vw format in parts also.
I believe the function should be modified for this to work, right? I can take a look at it and provide a solution later. 

This small example shows the idea, for point (2):

```
library(data.table)
N = 100000
ncls = 90
dt = data.table(rep(1,N))
lvls = c("tea", "coffee", "chocolate")
for(i in 2:ncls){
  tt = data.table(sample(lvls, N, replace = T))
  setnames(tt, colnames(tt), paste("V",i, sep = ""))
  dt = cbind(dt, tt)
}
namespaces = list(n = list(varName = sprintf('V%s', 2:ncls), keepSpace=F, dtype='numeric'))
write.table(dt, file='../data.csv', sep='|')

# So far it's just writing the data to disk. The idea is to read it using fread in chunks.
# Take this as pseudocode.

chunks = 10
rows = seq(0, N, N / chunks)
while(i < max(chunks)){
  # This would be reading the file in "chunks"
  dt = fread('../data.csv', skip = rows[i], nrows=rows[i+1])
  
  print(i)
  # I was thinking a useful feature for dt2vw woud be to add an 
  # option to write the file out in parts
  # perhaps adding an append=TRUE argument.
  dt2vw(data = dt, fileName = "../out.vw",
        namespaces = namespaces, target = "V1",
        weight = NULL, hard_parse = F)
  
}
```

**Comment by martinbel** [comment]
One question, related to the function/input format. When I define a feature as factor, it just outputs the value. This might not work if the factors are encoded as numbers (when there are several factor variables, all encoded as numbers). Isn't it better to encode categorical features as VariableName_value? 

`e.g` 
`-1 | NUM:1 VAR100_1 VAR110_1`

This way seems to be recommended here:
http://stackoverflow.com/questions/28640837/vowpal-wabbit-how-to-represent-categorical-features

I just copy the relevant part.
Here's the short cheat-sheet for encoding variables in vowpal wabbit:

```
Variable type       How to encode                readable example
-------------                 -------------                      ----------------
boolean             only encode the true case    is_alive
categorical         append value to name         color=green
ordinal+monotonic   :approx_value                education:2
numeric                     :actual_value                height:1.85
```

**Comment by SelimRaboudi** [comment]
Updated my PR to add the features you suggested. 

--> I added an argument append [boolean]. Makes it possible to process a file by chunks
--> Also for every character / factor variable, the name of the variable precedes its value, separated by "_".
Therefore even if you have a factor variable with numerical values (ex: 1->5), it will be written in the file as NameVar_value. This makes the script a bit slower (haven't optimized it yet) but it's worth it I believe.
